File: library-management/python/tests/recording_notifier.py
from __future__ import annotations

from dataclasses import dataclass

from library_management import Member


@dataclass
class Notification:
    member: Member
    message: str


class RecordingNotifier:
    def __init__(self) -> None:
        self._sent: list[Notification] = []

    @property
    def sent(self) -> list[Notification]:
        return list(self._sent)

    def send(self, member: Member, message: str) -> None:
        self._sent.append(Notification(member, message))

    def notifications_for(self, member: Member) -> list[Notification]:
        return [n for n in self._sent if n.member is member]

    def availability_notifications_for(self, member: Member) -> list[Notification]:
        return [n for n in self.notifications_for(member) if "available" in n.message]

    def expiration_notifications_for(self, member: Member) -> list[Notification]:
        return [n for n in self.notifications_for(member) if "expired" in n.message]

File: library-management/typescript/tests/RecordingNotifier.ts
import type { Member } from '../src/Member.js';
import type { Notifier } from '../src/Notifier.js';

export interface Notification { member: Member; message: string; }

export class RecordingNotifier implements Notifier {
  private _sent: Notification[] = [];
  get sent(): readonly Notification[] { return this._sent; }
  send(member: Member, message: string): void { this._sent.push({ member, message }); }
  notificationsFor(member: Member): Notification[] {
    return this._sent.filter((n) => n.member === member);
  }
  availabilityNotificationsFor(member: Member): Notification[] {
    return this.notificationsFor(member).filter((n) => n.message.includes('available'));
  }
  expirationNotificationsFor(member: Member): Notification[] {
    return this.notificationsFor(member).filter((n) => n.message.includes('expired'));
  }
}
